### PR TITLE
Allow specifying extra extensions to enable in Druid

### PR DESCRIPTION
**Pull request by soenkeliebau**
# Description

Currently the only way to enable extensions not enabled by default is to use config overrides.
This has the drawback that users need to specify all extensions to load, so basically anticipate what extensions the operator will enable and then add the ones they want.

This PR adds a field `additionalExtensions` to the clusterConfig - these extensions will be added to any extensions the operator deems are needed, so the user can just specify the extensions the care about here.

## Definition of Done Checklist

- Not all of these items are applicable to all PRs, the author should update this template to only leave the boxes in that are relevant
- Please make sure all these things are done and tick the boxes
 
```[tasklist]
# Author
- [ ] Changes are OpenShift compatible
- [ ] CRD changes approved
- [ ] Helm chart can be installed and deployed operator works
- [ ] Integration tests passed (for non trivial changes)
```

```[tasklist]
# Reviewer
- [ ] Code contains useful comments
- [ ] (Integration-)Test cases added
- [ ] Documentation added or updated
- [ ] Changelog updated
- [ ] Cargo.toml only contains references to git tags (not specific commits or branches)
```

```[tasklist]
# Acceptance
- [ ] Feature Tracker has been updated
- [ ] Proper release label has been added
```

Once the review is done, comment `bors r+` (or `bors merge`) to merge. [Further information](https://bors.tech/documentation/getting-started/#reviewing-pull-requests)

### CODE REVIEW

File: rust/operator-binary/src/extensions.rs [hunk 1]
```diff
@@ -38,5 +38,18 @@ pub fn get_extension_list(
         extensions.push(EXT_S3.to_string());
     }
 
+    // Add user specified extensions to the list of loaded extensions if any are present
+    for additional_extension in &druid.spec.cluster_config.additional_extensions {
+        if extensions.contains(additional_extension) {
```
Comment (sbernauer) [inline rust/operator-binary/src/extensions.rs:43]: O(n^2) here we come :) Just kidding, you could use a BTreeMap - feel free to ignore

**Review by sbernauer — COMMENTED** [review]
LGTM in principle.

Looking at e.g. https://github.com/orgs/stackabletech/discussions/15 I think we might want to add dedicated extensions in the future - in this case for google cloud extensions with bucket name and credentials secret.
The mechanism from this PR might co-exist, but I would like to model them together and not go with a Vec<String> for now and later e.g. realizing we want a BTreeMap<String, Whatever>

Let's wait for Arch meeting to discuss CRD change

**Comment by soenkeliebau** [comment]
> Looking at e.g. https://github.com/orgs/stackabletech/discussions/15 I think we might want to add dedicated extensions in the future - in this case for google cloud extensions with bucket name and credentials secret. The mechanism from this PR might co-exist, but I would like to model them together and not go with a Vec for now and later e.g. realizing we want a BTreeMap<String, Whatever>

Agree in principle, but for this specific example, wouldn't that be better handled in the deep storage config?

**Comment by sbernauer** [comment]
I think the gcs extension things includes more stuff such as ingesting data from gcs, not only deep storage

**Comment by fhennig** [comment]
Hey @sbernauer, I'm reviving this again because we had another user as for how to add extensions, I think it was the `avro` extension. It doesn't have any configuration, so a list would be perfectly fine. I'd really like to make this easier for users as it is currently quite easy to make a mistake when editing the list by hand in the overrides. The avro extension doesn't have any settings, so a list would work here.

Would you maybe like to expand on your design a bit? Maybe we can get something fleshed out to implement soonish.

I have also an alternative idea, which is to just load a lot more extensions by default - just thinking out loud here, I don't really know, are there any downsides to, let's say, load the avro extensions always? It's already in the image, so it's not about space.

**Comment by sbernauer** [comment]
Have not spend many thoughts on this, my only requirement is that we can add properties to extensions. Does not need to be right now, just give us the chance to add that non-breaking. For avro specifically I don't see a reason to not always enabled it.
CRD **could** look something like
```
kind: DruidCluster
spec:
  clusterConfig:
    additionalExtensions:
      google: # optional
        credentialsSecret: my-google-credentials # mandatory, gets put into GOOGLE_APPLICATION_CREDENTIALS
        bucket: my-bucket # mandatory
        prefix: / # mandatory, defaults to /
        maxListingLength: 1024 # optional
      avro: {} # optional
```

As a first step we could simply add all extensions that don't need any configuration options (such as avro)

**Comment by fhennig** [comment]
Okay yes seems reasonable and also still easy to use.

**Comment by labrenbe** [comment]
Closing because this is already implemented by #547.

**Pull request closed without merging**